polkadot/runtime/parachains/src/paras_inherent/mod.rs

```diff
@@ -1110,53 +1095,117 @@ fn limit_and_sanitize_disputes<
 	}
 }
 
-// Filters statements from disabled validators in `BackedCandidate`, non-scheduled candidates and
-// few more sanity checks. Returns `true` if at least one statement is removed and `false`
-// otherwise.
-fn filter_backed_statements_from_disabled_validators<T: shared::Config + scheduler::Config>(
-	backed_candidates_with_core: &mut Vec<(
-		BackedCandidate<<T as frame_system::Config>::Hash>,
-		CoreIndex,
-	)>,
+// Helper function for filtering candidates which don't pass the given predicate. When/if the first
+// candidate which failes the predicate is found, all the other candidates that follow are dropped.
+fn filter_candidates<
+	T: inclusion::Config + paras::Config + inclusion::Config,
+	F: FnMut(ParaId, &BackedCandidate<T::Hash>) -> bool,
+>(
+	candidates: &mut BTreeMap<ParaId, Vec<BackedCandidate<T::Hash>>>,
+	mut pred: F,
+) {
+	for (para_id, candidates) in candidates.iter_mut() {
+		let mut latest_valid_idx = None;
+
+		for (idx, candidate) in candidates.iter().enumerate() {
+			if pred(*para_id, candidate) {
+				// Found a valid candidate.
+				latest_valid_idx = Some(idx);
+			} else {
+				break
+			}
+		}
+
+		if let Some(latest_valid_idx) = latest_valid_idx {
+			candidates.truncate(latest_valid_idx + 1);
+		} else {
+			candidates.clear();
+		}
+	}
+
+	candidates.retain(|_, c| !c.is_empty());
+}
+
+// Helper function for filtering candidates which don't pass the given predicate. When/if the first
+// candidate which failes the predicate is found, all the other candidates that follow are dropped.
+fn filter_candidates_with_core<
+	T: inclusion::Config + paras::Config + inclusion::Config,
+	F: FnMut(ParaId, &mut BackedCandidate<T::Hash>, CoreIndex) -> bool,
+>(
+	candidates: &mut BTreeMap<ParaId, Vec<(BackedCandidate<T::Hash>, CoreIndex)>>,
+	mut pred: F,
```

Comment:
do we plan to reuse this function with diff predicate ? currently there is only one, so I'd vote for dropping it and renaming the function to a specialized name.

Reply:
yeah, I think I can make filter_candidates more generic and remove filter_candidates_with_core

Reply:
done